src/JsonSchema/SchemaFactory.php

```diff
@@ -210,15 +241,37 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
                 $className = $valueType->getClassName();
             }
 
-            $valueSchema = $this->typeFactory->getType(new Type($builtinType, $type->isNullable(), $className, $isCollection, $keyType, $valueType), $format, $propertyMetadata->isReadableLink(), $serializerContext, $schema);
+            $valueSchema[] = $this->typeFactory->getType(new Type($builtinType, $type->isNullable(), $className, $isCollection, $keyType, $valueType), $format, $propertyMetadata->isReadableLink(), $serializerContext, $schema);
         }
 
-        if (\array_key_exists('type', $propertySchema) && \array_key_exists('$ref', $valueSchema)) {
-            $propertySchema = new \ArrayObject($propertySchema);
-        } else {
-            $propertySchema = new \ArrayObject($propertySchema + $valueSchema);
+        $valueSchema = array_unique($valueSchema, \SORT_REGULAR);
```

Comment:
not a huge fan of the call, we could name our schemas and just try to `isset` before setting the value?